languages/rust/tests/parity.rs

```diff
@@ -0,0 +1,139 @@
+use oso::host::ToPolar;
+use oso::host::{Class, HostClass};
+use oso::Oso;
+
+macro_rules! res {
+    ($res:expr) => {
+        $res.unwrap().next().unwrap().unwrap();
+    };
+    (@not $res:expr) => {
+        assert!($res.unwrap().next().is_none());
+    };
+}
+
+#[derive(Clone)]
+struct A {
+    x: String,
+}
+
+impl HostClass for A {}
+
+impl A {
+    pub fn new(x: String) -> Self {
+        Self { x }
+    }
+
+    pub fn foo(&self) -> i32 {
+        -1
+    }
+}
+
+// pub trait A {}
+
+// TODO
+// struct D;
+
+// impl A for D {
+
+// }
+
+// oso.register_class(A)
+
+pub mod b {
+    #[derive(Clone, Default)]
+    pub struct C {
+        pub y: String,
+    }
+
+    impl C {
+        pub fn new(y: String) -> Self {
+            Self { y }
+        }
+
+        pub fn foo(&self) -> i32 {
+            -1
+        }
+    }
+
+    impl oso::host::HostClass for C {}
+}
+
+pub fn custom_c_constructor(y: String) -> b::C {
+    b::C::new(y)
+}
+
+#[test]
+fn test() {
+    let mut oso = Oso::new();
+    tracing_subscriber::fmt::init();
+
+    Class::with_constructor(A::new)
+        .name("A")
+        .add_attribute_getter("x", |a_self: &A| a_self.x.clone())
+        .add_method("foo", A::foo)
+        .register(&mut oso)
+        .unwrap();
```

Comment:
I really don't understand what's going on so this might be way off base, but why can't a lot of this configuration be done through the `HostClass` trait? 

Also, are we anticipating that classes would actually be registered this way? I find it more confusing to call `register()` on the class with `oso` as the arg, than the other way around.

Reply:
We can cover the implementation specifics after going through it. A couple of points on the API though:

- This is the long-form registration. Most would be handled by a proc macro along the lines of:
```
#[register_class(name="Foo")]
struct Foo { ... }
```
 - Maybe adding register as a method on `Class` was overkill. There are a couple of subtle things it's doing that makes it useful though